### PR TITLE
Implement test for internal/core/alogrithm/ngt/option and refactor

### PR DESCRIPTION
**Pull request by kevindiu**
Signed-off-by: kevindiu <kevin_diu@yahoo.com.hk>

### Description:

This PR implement the test for internal/core/alogrithm/ngt/option test.
We implement a new error option called ErrUnsetOption to print the debug log if this error returned.
I updated the coding guideline for above error.
It also refactor the option implementation to return InvalidOption/CriticalOption/UnsetOption error.

*Author review required due to logic changes

### Environment:

- Go Version: 1.16.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/core/algorithm/ngt/option.go [hunk 1]
```diff
@@ -104,58 +115,69 @@ func WithDistanceTypeByString(dt string) Option {
 		d = Hamming
 	case "cosine", "cos":
 		d = Cosine
-	case "normalizedangle", "normalized angle", "normalized ang", "nang", "nangle":
+	case "normalizedangle", "normalizedang", "nang", "nangle":
 		d = NormalizedAngle
-	case "normalizedcosine", "normalized cosine", "normalized cos", "ncos", "ncosine":
+	case "normalizedcosine", "normalizedcos", "ncos", "ncosine":
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option.go:120]: Could you please teach me about these changes? 🙏

File: internal/core/algorithm/ngt/option_test.go [hunk 2]
```diff
@@ -0,0 +1,1443 @@
...
+		{
+			name: "set success when the size is -100",
+			args: args{
+				size: -100,
+			},
+			want: want{
+				obj: &T{},
+				err: errors.New("invalid option, name: BulkInsertChunkSize, val: -100"),
+			},
+		},
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option_test.go:245]: should we add more test case like maxint,minint?

File: internal/core/algorithm/ngt/option_test.go [hunk 3]
```diff
@@ -0,0 +1,1443 @@
...
+		{
+			name: "return error when the size is 65537",
+			args: args{
+				size: 65537,
+			},
+			want: want{
+				obj: &T{},
+				err: errors.New("invalid critical option, name: dimension, val: 65537: dimension size 65537 is invalid, the supporting dimension size must be between 2 ~ 65536"),
+			},
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option_test.go:370]: same
https://github.com/vdaas/vald/pull/1251/files#r638391751

File: internal/core/algorithm/ngt/option_test.go [hunk 4]
```diff
@@ -0,0 +1,1443 @@
...
+		{
+			name: "return error when distance type is invalid",
+			args: args{
+				dt: "invalid type",
+			},
+			want: want{
+				err: errors.ErrUnsupportedDistanceType,
+				obj: &T{},
+			},
+		},
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option_test.go:580]: Could you please add test case of empty pattern?

File: internal/core/algorithm/ngt/option_test.go [hunk 5]
```diff
@@ -0,0 +1,1443 @@
...
+		{
+			name: "return error when object type is invalid",
+			args: args{
+				ot: "invalid",
+			},
+			want: want{
+				obj: &T{},
+				err: errors.New("invalid critical option, name: objectType, val: 0: unsupported ObjectType"),
+			},
+		},
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option_test.go:856]: same
https://github.com/vdaas/vald/pull/1251/files#r638392052

File: internal/core/algorithm/ngt/option.go (hunk 1 above)
Comment (kevindiu) [reply internal/core/algorithm/ngt/option.go:120]: due to the logic in line 107, the space character will be removed, so comparing with string contains space is no sense, so I removed it.

File: internal/core/algorithm/ngt/option_test.go [hunk 6]
```diff
@@ -0,0 +1,1504 @@
...
+			name: "return error when object type is empty string",
+			args: args{
+				ot: "",
+			},
+			want: want{
+				obj: &T{},
+				err: errors.New("invalid critical option, name: objectType, val: 0: unsupported ObjectType"),
+			},
+		},
+	}
```
Comment (hlts2) [inline internal/core/algorithm/ngt/option_test.go:918]: Could you please add test pattern that contains `-` and `_` ? 🙏

**Review by hlts2 — DISMISSED** [review]
LGTM

File: internal/core/algorithm/ngt/option.go (hunk 1 above)
Comment (vankichi) [reply internal/core/algorithm/ngt/option.go:120]: Did you add the test case about your changes?

Comment (kevindiu) [reply internal/core/algorithm/ngt/option.go:120]: Sorry I didn't add the test case for it. I will add new test case for it later.

File: internal/core/algorithm/ngt/option_test.go [hunk 7]
```diff
@@ -0,0 +1,1443 @@
...
+			name: "set success when the size is 100",
+			args: args{
+				size: 100,
+			},
+			want: want{
+				obj: func() *T {
+					t := &T{
+						dimension: 100,
+					}
+					t.setup()
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/core/algorithm/ngt/option_test.go:330]: plz fix it

File: internal/core/algorithm/ngt/option_test.go [hunk 8]
```diff
@@ -0,0 +1,1443 @@
...
+			args: args{
+				dt: "invalid type",
+			},
+			want: want{
+				err: errors.ErrUnsupportedDistanceType,
+				obj: &T{},
+			},
+		},
+	}
+	for _, tc := range tests {
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/core/algorithm/ngt/option_test.go:643]: plz fix it

[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/core/algorithm/ngt/option_test.go:646]: fixed

File: internal/core/algorithm/ngt/option.go [hunk 9]
```diff
@@ -104,58 +115,69 @@ func WithDistanceTypeByString(dt string) Option {
...
 		default:
-			return errors.ErrUnsupportedDistanceType
+			err := errors.ErrUnsupportedDistanceType
+			return errors.NewErrCriticalOption("distanceType", t, err)
 		}
 		return nil
 	}
 }
 
+// WithObjectTypeByString represents the option to set the object type for NGT.
```
Comment (vankichi) [inline internal/core/algorithm/ngt/option.go:180]:
```suggestion
// WithObjectTypeByString represents the option to set the object type by string for NGT.
```
It would be better differ from the explanation of `WithObjectType` 🤔 
L192

**Review by vankichi — COMMENTED** [review]
Overall LGTM

**Review by hlts2 — APPROVED** [review]

**Comment by kevindiu** [comment]
@kpango @rinx Since this PR includes changes on functional option handling, can you please review this PR when you have time?

File: internal/core/algorithm/ngt/option.go [hunk 10]
```diff
@@ -51,38 +52,47 @@ var (
...
 		return nil
 	}
 }
 
+// WithIndexPath represents the option to set the index path for NGT.
 func WithIndexPath(path string) Option {
 	return func(n *ngt) error {
 		if len(path) == 0 {
-			return nil
+			return errors.NewErrInvalidOption("indexPath", path)
```
Comment (rinx) [inline internal/core/algorithm/ngt/option.go:67]: I think running agents without `indexPath` is normal usage.
Does it show a warn error in log?
If it does, i think it is better to create a new level of option error like `ErrUnsetOption` (maybe there're better names) and it causes only a debug log.

Comment (kevindiu) [reply internal/core/algorithm/ngt/option.go:67]: Yes, it shows the warn log in agent.
@vankichi @hlts2 what do you think?

Comment (hlts2) [reply internal/core/algorithm/ngt/option.go:67]: thank you for comment.

>  i think it is better to create a new level of option error like ErrUnsetOption (maybe there're better names) and it causes only a debug log.

I understood it. and I think it looks good to me.
What about the `IgnoreOption` for the option error name?

Comment (kevindiu) [reply internal/core/algorithm/ngt/option.go]: @hlts2 I have fixed rinx comment, can you please review it again?

Comment (kevindiu) [reply internal/core/algorithm/ngt/option.go]: @hlts2 Sorry for late comment.
I think `IgnoreOption` like the option is being ignored, which is not (we will print the debug log), so I think `ErrUnsetOption` is better for me.
What do you think?

**Comment by vankichi** [comment]
Overall LGTM.
Please add the test to cover codes you added in this PR.
https://github.com/vdaas/vald/pull/1251/commits/c8ddf3bbaea3b4135bda5f522c1e5de086424064

**Review by vankichi — APPROVED** [review]

**Review by hlts2 — APPROVED** [review]

**Comment by kevindiu** [comment]
@rinx @kpango we have fixed the comment about the `ErrUnsetOption`, can you please review this PR again?

**Review by rinx — APPROVED** [review]

File: internal/errors/option.go [hunk 11]
```diff
@@ -105,3 +105,48 @@ func (e *ErrCriticalOption) Error() string {
 func (e *ErrCriticalOption) Unwrap() error {
 	return e.origin
 }
+
+// ErrUnsetOption represents the unset option error.
+type ErrUnsetOption struct {
```
Comment (kpango) [inline internal/errors/option.go:110]: ErrIgnoredOption sounds more understandable

Comment (kevindiu) [reply internal/errors/option.go:110]: @kpango I think Ignore means it is ignored, and will not process anything, so I think ErrUnsetOption is better.
What do you think?

Comment (kpango) [reply internal/errors/option.go:110]: @kevindiu but for IndexPath it seems ignored...

Comment (kpango) [reply internal/errors/option.go:110]: How about ErrUnconfiguredOption?
I think the Unset is usually misunderstood as the meaning of the Getter/Setter pattern, which reverts the set to Empty, so I think it's better to use other words like Ignored or Unconfigured.

Comment (kevindiu) [reply internal/errors/option.go:110]: I start thinking that ErrIgnoreOption is reasonable that it is ignored and debug log will be printed.
@rinx @hlts2 @vankichi what do you think?

Comment (vankichi) [reply internal/errors/option.go:110]: 🤔

Comment (vankichi) [reply internal/errors/option.go:110]: IMO, `ErrIgnoredOption` is meaningful

Comment (kpango) [reply internal/errors/option.go:110]: @vankichi sounds great.

@kevindiu can you please fix it?

Comment (kevindiu) [reply internal/errors/option.go]: @kpango I have fixed this, thank you for your comment :)

Comment (hlts2) [reply internal/errors/option.go]: > what do you think?
ok,  sounds good.!
thank you

**Review by kpango — APPROVED** [review]
LGTM
